src/SonarScanner.MSBuild.PreProcessor/WebService/SonarWebService.cs

```diff
@@ -29,34 +29,43 @@
 using SonarScanner.MSBuild.PreProcessor.Protobuf;
 using SonarScanner.MSBuild.PreProcessor.Roslyn.Model;
 
-namespace SonarScanner.MSBuild.PreProcessor
+namespace SonarScanner.MSBuild.PreProcessor.WebService
 {
-    public sealed class SonarWebService : ISonarWebService
+    public abstract class SonarWebService : ISonarWebService
     {
         private const string OldDefaultProjectTestPattern = @"[^\\]*test[^\\]*$";
         private const string TestProjectPattern = "sonar.cs.msbuild.testProjectPattern";
+
+        protected readonly IDownloader downloader;
+        protected readonly Version serverVersion;
+        protected readonly ILogger logger;
         private readonly Uri serverUri;
-        private readonly IDownloader downloader;
-        private readonly ILogger logger;
-        private Version serverVersion;
 
-        public SonarWebService(IDownloader downloader, string server, ILogger logger)
-        {
-            Contract.ThrowIfNullOrWhitespace(server, nameof(server));
+        private bool disposed;
 
+        public Version ServerVersion => serverVersion;
```

Comment:
I do agree with you and I think this should be handled but not in this PR. We will consider this for one of the following ones.
I added a point here: https://github.com/SonarSource/sonar-scanner-msbuild/issues/1464

Reply:
This one I would like to do here already.

Reply:
I still prefer to do this later. `AnalysisConfigGenerator.GenerateFile(` is also using this and it could get more complicated.
I prefer to choose the battles and focus on what is important now.